File: a013(roman_numerals)/C++/roman_numerals.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

class Pair {
public:
  Pair() {}

  Pair(std::string key, int value):key(key),value(value) {}

  Pair(Pair const & p):key(p.key),value(p.value) {}

  int & getValue() {
    return value;
  }

  std::string getKey() const {
    return key;
  }

private:
  std::string key;
  int value;
};

class Dict {
public:
  Dict() {}

  void append(Pair pair) {
    pairs.push_back(pair);
  }

  Dict(Dict const & d):pairs(d.pairs) {}

  int & operator [] (std::string key) {
    for (Pair & pair: pairs) {
      if (pair.getKey() == key) {
        return pair.getValue();
      }
    }
    throw std::string("Key doesn't exist.");
  }

  std::vector<Pair> & getPairs() {
    return pairs;
  }

private:
  std::vector<Pair> pairs;
};

Dict roman_dict;

int roman_to_decimal(std::string roman_number) {
  int deciml_number = 0, i = 0;
  while (i < roman_number.length()) {
    if (i + 1 < roman_number.length()) {
      if(roman_number[i] == 'I' && roman_number[i + 1] == 'V' || roman_number[i] == 'I' && roman_number[i + 1] == 'X'
      || roman_number[i] == 'X' && roman_number[i + 1] == 'L' || roman_number[i] == 'X' && roman_number[i + 1] == 'C'
      || roman_number[i] == 'C' && roman_number[i + 1] == 'D' || roman_number[i] == 'C' && roman_number[i + 1] == 'M') {
        deciml_number += roman_dict[std::string(1, roman_number[i]) + std::string(1, roman_number[i + 1])];
        i += 2;
        continue;
      }
    }
    deciml_number += roman_dict[std::string(1, roman_number[i])];
    i++;
  }
  return deciml_number;
}

std::string decimal_to_roman(int &decimal_number) {
  std::sort(roman_dict.getPairs().begin(), roman_dict.getPairs().end(), [](Pair former, Pair latter) {
    return former.getValue() > latter.getValue();
  });
  if (decimal_number == 0) return "ZERO";
  std::string roman_number = "";
  for (Pair p: roman_dict.getPairs()) {
    int times = decimal_number / p.getValue();
    decimal_number %= p.getValue();
    while (times != 0) {
      roman_number += p.getKey();
      times--;
    }
  }
  return roman_number;
}

int main(void) {
  
  roman_dict.append(Pair("M", 1000));
  roman_dict.append(Pair("CM", 900));
  roman_dict.append(Pair("D", 500));
  roman_dict.append(Pair("CD", 400));
  roman_dict.append(Pair("C", 100));
  roman_dict.append(Pair("XC", 90));
  roman_dict.append(Pair("L", 50));
  roman_dict.append(Pair("XL", 40));
  roman_dict.append(Pair("X", 10));
  roman_dict.append(Pair("IX", 9));
  roman_dict.append(Pair("V", 5));
  roman_dict.append(Pair("IV", 4));
  roman_dict.append(Pair("I", 1));

  std::string r1, r2;
  while (std::cin >> r1 >> r2) {
    int n1 = roman_to_decimal(r1);
    int n2 = roman_to_decimal(r2);
    int difference = abs(n1 - n2);
    std::cout << decimal_to_roman(difference) << '\n';
  }

  return 0;
}

File: a013(roman_numerals)/Python/roman_numerals.py
roman_dict = {'M': 1000, 'CM': 900, 'D': 500, 'CD': 400, 'C': 100, 'XC': 90, 'L': 50,
              'XL': 40, 'X': 10, 'IX': 9, 'V': 5, 'IV': 4, 'I': 1}

def roman_to_decimal(roman_number: str) -> int:
    decimal_number, i = 0, 0
    while i < len(roman_number):
        if i + 1 < len(roman_number):
            check = roman_number[i] + roman_number[i + 1]
            if check == 'IV' or check == 'IX' or check == 'XL' \
            or check == 'XC' or check ==  'CD' or check == 'CM':
                decimal_number += roman_dict[roman_number[i:i+2]]
                i += 2
                continue
        decimal_number += roman_dict[roman_number[i]]
        i += 1
    return decimal_number

def decimal_to_roman(decimal_number: int) -> str:
    if decimal_number == 0: return 'ZERO'
    roman_number = []
    for numeral, value in roman_dict.items():
        times = decimal_number // value;
        roman_number.append(numeral * times);
        decimal_number %= value;
    return ''.join(roman_number)


while True:
    try:
        r1, r2 = input().split()
        n1 = roman_to_decimal(r1)
        n2 = roman_to_decimal(r2)
        difference = abs(n1 - n2)
        print(decimal_to_roman(difference))
    except:
        break
